Improve tox, Travis, and local test / build configs

coalkettler opened the pull request:
## Overview

This introduces a variety of local development fixes, Travis / tox optimizations and best practices, and drops support for end-of-life Python and Django versions.

Most of these updates reflect similar (and massively informative 💖) contributions made in https://github.com/azavea/django-amazon-ses/compare/3.0.0...96e53398aee537feab118c39312b40f22e5f390e.

* Drop support for Python 2.7 and 3.4
* Drop support for Django 2.0 and 2.1
* Fix Django `master` builds which would have previously resolved to a 3.x version
* Fix formatting / linting configs for `scripts/test`
* Fix source code formatting issues reported by `black`
* Move most settings from `settings.py` to `settings.cfg`
* Remove deprecated config options from `settings.py` / `settings.cfg`
* Remove redundant config options from `tox.ini` and `travis.yml`
* Run linting in separate `tox` environment so that it isn't run unnecessarily on every build

Closes #15 

### Notes

There's some formatting fixes here. You'll probably want to hide whitespace diffs.

## Testing Instructions

 * Review [Travis build logs](https://travis-ci.org/azavea/django-ecsmanage/builds/655979973). Verify that `TOXENV=lint` job runs.
 * Run `scripts/update` and `scripts/test` locally and confirm that all linting and tests pass.

coalkettler committed:
Drop support for Python 2.7, 3.4

coalkettler committed:
Remove deprecated license_file option from setup.cfg

LICENSE file will be included in dist automatically

coalkettler committed:
Ignore eggs when linting in test script

coalkettler committed:
Provide path to Black linter command

coalkettler committed:
Fix formatting errors reported by Black

coalkettler committed:
Remove deprecated pytest-runner and pytest alias

coalkettler committed:
Remove support for Django 2.0, 2.1

coalkettler committed:
Remove upper constraint on Django in install_requires

Allows testing the master branch of the Django repository

coalkettler committed:
Move setup.py configuration to setup.cfg

coalkettler committed:
Remove redundant flake8 config

Removed config options are already set as defaults or not relevant during a tox run

coalkettler committed:
Remove deprecated sudo option from Travis config, use default Ubuntu Xenial distribution

coalkettler committed:
Run linting in separate Tox environment only

coalkettler committed:
Update CHANGELOG with tox, Travis, and linting changes

coalkettler committed:
Use Python 3.8 for linting job

3.5 was being used which does not support f-strings without including the future-fstrings dependency

coalkettler commented on .travis.yml:
```diff
@@ -9,13 +7,16 @@ env:
+matrix:
+  include:
+    - env: TOXENV=lint
+      python: 3.8
```
The Python version is specified explicitly here 'cause it was running at 3.5 by default, which does not support f-strings, which we've patched over by installing `future-fstrings`. However `skip_install` in the `tox` config is enabled for this linting environment so we weren't actually installing it.

rbreslow replied:
It looks like we're still building 3.5 though and it's passing. The build output looks the same as 3.8. Since this isn't defined in `django-amazon-ses`, I'd rather remove it to achieve alignment.

```suggestion
```

coalkettler replied:
Yup, the 3.5 build installs `future-fstrings` as part of its [standard installation requirements](https://github.com/azavea/django-ecsmanage/blob/08b0bd72bbd212765960cd2e806ac2b00988979f/setup.cfg#L29-L32). So that's getting installed when we run unit tests for 3.5 with `tox` but not when we run linting.

`django-amazon-ses` is running linting with 3.5 (I think it defaults to whichever version is declared first in the `python` list in `travis.yml`) and [passing](https://travis-ci.org/azavea/django-amazon-ses/jobs/648061400) because it isn't using f-strings.

The `lint` `tox` environment was [failing](https://travis-ci.org/azavea/django-ecsmanage/jobs/655577224) when running on 3.5 and [passes](https://travis-ci.org/azavea/django-ecsmanage/jobs/657367558) on 3.8. We could install `future-fstrings` in the `tox` dependencies but I think it's preferable for us to lint with the most recent stable version of Python. That way we can use more recent versions of our linters and tools that might not support older versions of Python.

If anything I think we should update `django-amazon-ses` to explicitly use 3.8 unless we have made a conscious decision to use 3.5.

rbreslow replied:
If we're supporting 3.5 as the oldest version, I have a desire to make sure all of our checks pass on 3.5+. With that logic, what do you think of running linting across all versions? Can we install `future-fstring` and invoke linting in each build?

coalkettler replied:
I was basing these changes off of azavea/django-amazon-ses#61 which was introduced to stop linting across all versions. The rationale for that PR was that linting only needs to be run once for the whole project and we don't need to lint for each version. This speeds up the rest of the build by removing the need to lint on each variant build.

My observation was that the failure on 3.5 wasn't indicating a linting error in the source but a parsing problem with the linter's runtime.

rbreslow replied:
There are two rationales in my head:

- Support 3.8 so we don't have to install `future-fstrings`
- Support 3.5 so we know there is compatibility with the oldest, supported version

I don't feel strongly about either of them. I'm fine with you forging the path from here.

coalkettler replied:
I thought about this pretty hard, and I'm in favor of running our linter on 3.5.

We can catch syntax errors this way, but it feels kinda like blanket code coverage and something that we should be catching with automated tests. However we don't have code coverage on those paths right now and we'd need to prioritize increasing coverage.

Practicality beats principle for me so I'd rather we rely on a blanket approach than potentially miss errors that break backwards compatibility. I've confirmed that the install / test process would totally miss if we forgot to install `future-fstrings` and that's reason enough for me.

I'll install `future-fstrings` in the `lint` environment and get this merged up!

coalkettler commented on ecsmanage/management/commands/ecsmanage.py:
```diff
@@ -56,7 +56,7 @@ def handle(self, *args, **options):
             f"/clusters/{cluster_name}/tasks/{task_id}/details"  # NOQA
         )
 
-        self.stdout.write(self.style.SUCCESS(f"Task started! View here:\n{url}"))  # NOQA
```
The linter wasn't actually complaining about this so I removed the `#NOQA`.

coalkettler commented on ecsmanage/management/commands/ecsmanage.py:
```diff
@@ -113,7 +113,7 @@ def parse_response(self, response, key, idx=None):
                 try:
                     return response[key][0]
                 except (IndexError, TypeError):
-                    msg = f"Unexpected value for '{key}' in response: " f"{response}"  # NOQA
```
The linter wasn't actually complaining about this so I removed the `#NOQA`.

coalkettler commented on scripts/test:
```diff
@@ -26,9 +26,9 @@ function run_linters() {
+    ./.venv/bin/flake8 --exclude "*.pyc,.eggs,.venv"
     if command -v ./.venv/bin/black > /dev/null; then
-        ./.venv/bin/black --check --target-version py36
+        ./.venv/bin/black --check --diff --target-version py36 .
```
`black` wasn't actually working because it wasn't given a path. I've set it to check the whole repository and output the diff, like we do [here](https://github.com/azavea/stac-packer/blob/2286e62c4b7d4a35271698a2f337f5f345418e63/scripts/test#L29).

rbreslow replied:
Was it returning a status code but not actually checking anything?

coalkettler replied:
Yup! It just said it had nothing to do and returned `0`.

hectcastro commented:
I just reassigned this PR to @rbreslow because I think there is little chance that I'll be able to look at it before tomorrow afternoon. The only thing I wanted to note is that a few more changes went into the `django-amazon-ses` module since 3.0.2. (I didn't compare them relative to the changes here, so it is possible that they're already been incorporated.)

See: https://github.com/azavea/django-amazon-ses/compare/3.0.0...develop

ColeKettler commented:
@hectcastro Handing this off to you since you reviewed the PRs that went into https://github.com/azavea/django-amazon-ses/compare/3.0.0...3.0.2. Almost all of the rationale for these changes is identical here. Happy to annotate the PR with those rationales too if it's helpful for posterity's sake!

coalkettler commented:
@hectcastro Sounds good! I saw those changes and also incorporated them. Pretty much everything relevant from  azavea/django-amazon-ses#52 through  azavea/django-amazon-ses#64 is reflected here.

@rbreslow The rationales for these changes are detailed in the PR range I just mentioned, but please let me know if you'd like me to annotate the changeset here for sake of clarity!

coalkettler commented on .travis.yml:
```diff
@@ -1,5 +1,3 @@
-dist: xenial
```
This is the default for Travis builds, setting it here isn't necessary. (azavea/django-amazon-ses#60)

coalkettler commented on .travis.yml:
```diff
@@ -1,5 +1,3 @@
-dist: xenial
-sudo: false
```
This option has been deprecated by Travis. (azavea/django-amazon-ses#60)

coalkettler commented on setup.cfg:
```diff
@@ -1,8 +1,39 @@
-[bdist_wheel]
-universal = 1
```
Since we've dropped support for Python 2, this isn't actually a universal wheel anymore. (azavea/django-amazon-ses#53)

rbreslow replied:
https://pythonwheels.com/#pure-wheel 👍

coalkettler commented on setup.cfg:
```diff
@@ -1,8 +1,39 @@
-universal = 1
-
 [metadata]
-license_file = LICENSE
```
This option has been deprecated. It's been replaced by `license_files`, but the `LICENSE` file will be automatically included in the distribution so it's not necessary here. (azavea/django-amazon-ses#54)

coalkettler commented on setup.cfg:
```diff
@@ -1,8 +1,39 @@
+setup_requires =
+    setuptools_scm ==3.*
 
-[aliases]
-test = pytest
```
We're not actually using `pytest`, we're relying on Django's test utilities.

coalkettler commented on setup.py:
```diff
@@ -1,53 +1,7 @@
-        "Programming Language :: Python :: 3.8",
-    ],
-)
+setup(use_scm_version=True,)
```
This option [needs to be included](https://github.com/pypa/setuptools_scm#setupcfg-usage) in `setup.py` rather than `setup.cfg`.

rbreslow replied:
Unfortunate, but 👍

coalkettler commented on tox.ini:
```diff
@@ -11,24 +10,22 @@ envlist =
     {envpython} setup.py check -m -r -s
     flake8 .
-    django-admin test --noinput
+skip_install = true
```
We don't actually need to install the django-ecsmanage package and dependencies in order to lint it. This saves some time during this phase of the build. (azavea/django-amazon-ses#61)

coalkettler commented on tox.ini:
```diff
@@ -1,8 +1,7 @@
-       py{27,34,35,36,37}-django111
-       py{34,35,36,37}-django20
-       py{35,36,37}-django21
+       lint
```
We can run the linting once in a separate `tox` environment instead of rerunning it for every environment. This cuts down on build time. (azavea/django-amazon-ses#61)

coalkettler commented on tox.ini:
```diff
@@ -11,24 +10,22 @@ envlist =
 
 [flake8]
 max-line-length = 88
-exclude = .tox,*.egg,build,data
-select = E,W,F
```
These options are already set as `flake8` defaults. (azavea/django-amazon-ses#57)

coalkettler commented on setup.py:
```diff
@@ -1,53 +1,7 @@
-    author="Azavea, Inc.",
-    author_email="systems@azavea.com",
-    install_requires=[
-        "Django>=1.11,<3.1",
```
The upper bound of this was relaxed in `setup.cfg` to allow building the Django `master` branch. (azavea/django-amazon-ses#63)

coalkettler commented on setup.cfg:
```diff
@@ -1,8 +1,39 @@
-
 [metadata]
-license_file = LICENSE
+name = django-ecsmanage
```
Moving these settings from `setup.py` to `setup.cfg` allows for a more declarative, config file-style approach. (azavea/django-amazon-ses#56)

coalkettler committed:
Link to relevant PRs in CHANGELOG

coalkettler committed:
Fix indentation in tox config

coalkettler commented:
@rbreslow I thought about this more and realized it would be clearer to annotate the changeset with references to the corresponding PRs in [azavea/django-amazon-ses](https://github.com/azavea/django-amazon-ses). All set for review!

rbreslow commented on setup.py:
```diff
@@ -1,53 +1,7 @@
 
 # allow setup.py to be run from any path
 os.chdir(os.path.normpath(os.path.join(os.path.abspath(__file__), os.pardir)))
 
```
Can we do this to align with https://github.com/azavea/django-amazon-ses/blob/develop/setup.py?

```suggestion
```

rbreslow commented on setup.cfg:
```diff
@@ -1,8 +1,39 @@
+[options.extras_require]
+tests =
+    flake8 >=3.7.7
+    black; python_version > "3.6"
```
I'm interested to know more about what this line does. We're building 3.5, but this suggests that a component of our application is only compatible with 3.6+.

@hectcastro do you have more context?

coalkettler replied:
If somebody chooses to install the optional test dependencies (`pip install -e ".[tests]"`) this will install `black` only if they're using Python 3.6 or greater. It's excluded from the standard install.

This is relevant in `scripts/update` but we don't actually rely on `extras_require` in the Travis / `tox` build, and we aren't actually running `black` as part of the Travis build. Could be a good idea to do that, I've created #17 to track it.

hectcastro replied:
I think that this constraint was added to appease the Python version constraints defined by Black itself:

> It requires Python 3.6.0+ to run but you can reformat Python 2 code with it, too.

rbreslow reviewed (approved):
I appreciate the level of detail and consistency in this PR. Left a few comments, but you can merge at your discretion. 👍

coalkettler committed:
Remove working directory change from setup.py

coalkettler commented:
@rbreslow I'd like to get your feedback on explicitly running linting in 3.8 vs. 3.5 prior to merging. I think there's an opportunity to set precedent here for other projects and I don't want to lose the thread of it. Back to you!

coalkettler committed:
Run linter on Python 3.5

This will help catch syntax that breaks backwards compatibility with
the oldest version of Python we support. flake8 will parse the AST and
throw a syntax error if it encounters a problem.

coalkettler committed:
Ignore lint errors on lines that are processed by future_fstrings

coalkettler commented:
@rbreslow Oh goodness, turns out that `future-fstrings` transforms the source, so we're linting processed code and turning up errors when the original is fine. 😷 There are only three lines where this is a problem so I've disabled linting for those lines and changed formatting to satisfy `black`. There were `NOQA` directives on most of those lines originally, my guess is this has already been recognized by somebody who was running linting locally with < 3.6.

I don't love that solution but it's gonna be a moot point in 6 - 7 months anyway when we can drop support for 3.5 and uninstall `future-fstrings`. I've created #18 to keep track of this context when we eventually drop 3.5.
